relentless/core/variable.py

```diff
@@ -0,0 +1,550 @@
+        return self._params
+
+    def dependency_graph(self):
+        """Constructs a networkx graph of all variable objects on which the specified DependentVariable depends.
+
+        The graph nodes are all variable objects which DependentVariable are dependent on.
+        The directed graph edges represent dependencies as connections between variables.
+        The graph can also take into account the dependencies of a DependentVariable which has
+        multiple parameters as the same object.
+
+        Returns
+        -------
+        :py:class:`networkx.DiGraph`
+            The graph of all dependencies for this DependentVariable.
+
+        """
+        # construct graph of variables
+
+        # discover all variables (nodes) by quasi-depth first search
+        g = nx.DiGraph()
+        stack = [self]
+        visited = set()
+        while stack:
+            a = stack.pop()
+            if a not in visited:
+                g.add_node(a)
+                visited.add(a)
+                if isinstance(a, DependentVariable):
+                    for p,b in a.depends:
+                        stack.append(b)
+                        # add new edge for p if not already present, otherwise append to params list
+                        if (a,b) not in g.edges:
+                            g.add_edge(a,b,params=[p])
+                        else:
+                            g.edges[a,b]['params'].append(p)
+
+        return g
+
+    def derivative(self, var):
+        """Calculates the derivative of a DependentVariable object
+        with respect to another Variable object.
+
+        Parameters
+        ----------
+        var : :py:class:`Variable`
+            The variable with respect to which to take the derivative.
+
+        Returns
+        -------
+        float
+            The calculated derivative value.
+
+        Raises
+        ------
+        RuntimeError
+            If the specified DependentVariable object has any circular dependencies.
+
+        """
+        # if var is this variable, the derivative is trivially 1.0
+        if var is self:
+            return 1.0
+
+        # get dependency graph
+        g = self._assert_acyclic(self.dependency_graph())
+
+        # if var is not in the graph, then its derivative is trivially 0.0
+        if var not in g:
+            return 0.0
+
+        # add sum of parameter derivatives to edges between objects
+        for a,b,params in g.edges.data('params'):
+            g.edges[a,b]['deriv'] = np.sum([a._derivative(p) for p in params])
+
+        # compute chain rule along all paths to the variable
+        deriv = 0.
+        paths = nx.all_simple_paths(g, source=self, target=var)
+        for path in map(nx.utils.pairwise, paths):
+            deriv += np.prod([g.edges[edge]['deriv'] for edge in path])
+        return deriv
+
+    @abc.abstractmethod
+    def _derivative(self, param):
+        pass
+
+    def _assert_variable(self, v):
```

Comment:
Let's make this a classmethod, since it isn't specific to an instance:
```suggestion
    @classmethod
    def _assert_variable(cls, v):
```